usr/src/uts/common/fs/zfs/arc.c

```diff
@@ -1931,15 +2323,90 @@ arc_buf_try_copy_decompressed_data(arc_buf_t *buf)
  * the correct-sized data buffer.
  */
 static int
-arc_buf_fill(arc_buf_t *buf, boolean_t compressed)
+arc_buf_fill(arc_buf_t *buf, spa_t *spa, const zbookmark_phys_t *zb,
+    arc_fill_flags_t flags)
 {
+	int error = 0;
 	arc_buf_hdr_t *hdr = buf->b_hdr;
-	boolean_t hdr_compressed = (HDR_GET_COMPRESS(hdr) != ZIO_COMPRESS_OFF);
+	boolean_t hdr_compressed =
+	    (arc_hdr_get_compress(hdr) != ZIO_COMPRESS_OFF);
+	boolean_t compressed = (flags & ARC_FILL_COMPRESSED) != 0;
+	boolean_t encrypted = (flags & ARC_FILL_ENCRYPTED) != 0;
 	dmu_object_byteswap_t bswap = hdr->b_l1hdr.b_byteswap;
+	kmutex_t *hash_lock = (flags & ARC_FILL_LOCKED) ? NULL : HDR_LOCK(hdr);
 
 	ASSERT3P(buf->b_data, !=, NULL);
-	IMPLY(compressed, hdr_compressed);
+	IMPLY(compressed, hdr_compressed || ARC_BUF_ENCRYPTED(buf));
 	IMPLY(compressed, ARC_BUF_COMPRESSED(buf));
+	IMPLY(encrypted, HDR_ENCRYPTED(hdr));
+	IMPLY(encrypted, ARC_BUF_ENCRYPTED(buf));
+	IMPLY(encrypted, ARC_BUF_COMPRESSED(buf));
+	IMPLY(encrypted, !ARC_BUF_SHARED(buf));
+
+	/*
+	 * If the caller wanted encrypted data we just need to copy it from
+	 * b_rabd and potentially byteswap it. We won't be able to do any
+	 * further transforms on it.
+	 */
+	if (encrypted) {
+		ASSERT(HDR_HAS_RABD(hdr));
+		abd_copy_to_buf(buf->b_data, hdr->b_crypt_hdr.b_rabd,
+		    HDR_GET_PSIZE(hdr));
```

Comment:
Possibly dump question as I try to get my head wrapped around more of ZFS -- if `encrypted` is set (i.e. we take this branch), does that imply `HDR_GET_PSIZE(hdr) == HDR_GET_LSIZE(hdr)` -- if not, won't it mismatch when we goto byteswap?

Reply:
> does that imply HDR_GET_PSIZE(hdr) == HDR_GET_LSIZE(hdr)

`HDR_GET_PSIZE(hdr) == HDR_GET_LSIZE(hdr)` is not implied to be true. `b_rabd` is always implied to be `HDR_GET_PSIZE()` bytes large.  This might be a bug. We did some testing against big endian machines and I can say that at one point this code did seem to work. However, our options for testing machines were a PPC Macbook from 2003 and emulated PPC64 via QEMU. These we're both ridiculously slow and hard to work with so it wouldn't surprise me if this slipped through the cracks.

Reply:
i have SPARC64 with dilos and can import changes and we can try to test it.
how is stable now this update?